### DISCUSSION TITLE
Conditional checkbox for roles based system

### DISCUSSION THREAD

**Discussion by mohdaftab** [livewire/livewire#1870] [1. Help] [answered]
Hello,

I am working on a roles based system for an admin account which selects user role from select option.
I have a few roles in the list which can be multi selected and if any of them is selected I need all other roles which are not allowed for multiple roles to be unchecked and if there is a non-multiple role selected then I need all others to be unchecked.

I managed to do this using Javascript but as soon as livewire validation triggers on other fields all the checkboxes which I clicked on gets checked.

I got this fixed using debounce.50000ms on role select option but now the issue is I need some roles to emit a function to load users list in a dropdown with selected role but due to debounce it doesn't work until 50 seconds are gone. Please let me know what is the best way to work this around.

I am a beginner Livewire user and AlpineJS as well.
I also tried AlpineJS and tried to call a javascript function but it behaved the same way.

Please help me find the solution for this issue.

Thank you

**This is the javascript I am using**

`$(document).on("change", "input.user-role", function (e) {

if ($(this).is(":checked")) {
    role = $(this).val();

    if (
        role == "user" ||
        role == "team-manager" ||
        role == "service-provider" ||
        role == "service-associate" ||
        role == "crew" ||
        role == "vendor" ||
        role == "sales-associate" ||
        role == "review-associate"
        ) {
        $("input.user-role").each(function () {
            $(this).prop("checked",false);
        });    
        $("input.user-role[value='"+role+"']").prop("checked",true);
    } else if (
        role == "service-manager" ||
        role == "review-manager" ||
        role == "sales-manager"
    ) {
        $("input.user-role").each(function () {
            if (
                $(this).val() != "sales-manager" &&
                $(this).val() != "service-manager" &&
                $(this).val() != "review-manager"
            ) {
                $(this).prop("checked",false);
            }
        });    
        $("input.user-role[value='"+role+"']").prop("checked",true);

    }

    
}
});`

**And this is my blade template:**

`Add User Form

            <div class="form-row">
                @if (session()->has('message'))
                    <div class="col-12">
                        <div class="alert alert-success">
                            {{ session('message') }}
                        </div>
                    </div>
                @endif

                <div class="col-6">
                    <div class="form-group">
                        <label for="name">Name</label>
                        <input type="text" class="form-control" name="name" id="name" wire:model.lazy="name" aria-describedby="name" placeholder="Name">
                        @error('name') <div class="invalid-feedback d-inline">{{ $message }}</div> @enderror

                    </div>
                </div>

                <div class="col-6">
                    <div class="form-group">
                        <label for="email">Email</label>
                        <input type="email" class="form-control" name="email" id="email" wire:model.lazy="email" aria-describedby="email" placeholder="Email">
                        @error('email') <div class="invalid-feedback d-inline">{{ $message }}</div> @enderror

                    </div>
                </div>

                <div class="col-6">
                    <div class="form-group">
                        <label for="role">Choose Role</label>
                        
                        @foreach (getDefinedRoles() as $drolekey=>$drole)
                            @if (auth()->user()->isAllowed($drole))
                            <div class="custom-control custom-switch">
                                <input type="checkbox" class="custom-control-input user-role" wire:model.debounce.50000ms="role.{{$drole}}" id="role-{{ $drole }}" value="{{ $drole }}">
                                <label for="role-{{ $drole }}" class="custom-control-label">{{ ucwords(str_replace('-',' ',$drole)) }}</label>
                            </div>
                            @endif
                        @endforeach

                        @error('role') <div class="invalid-feedback d-inline">{{ $message }}</div> @enderror
                    </div>

                </div>

                <div class="col6">
                    {{ $parent_role }}
                </div>
            </div>

        </div>
        <div class="card-footer text-muted d-flex justify-content-between">
            <button type="submit" class="btn btn-dark" wire:click="submit">Add</button>
            <button type="reset" class="btn btn-danger">Reset</button>
        </div>
    </div>
</form>`

**Comment by iksaku** [comment, marked as answer]
I'm not sure if using JQuery is the best option for you. As you already discovered, working with proxied values is kind of hard and weird some times. 

Now, for your issue, remember that each time that you "check" one of your role options, it'll fire a Livewire request, which you  could use in the backend to automatically enable/disable or check/uncheck values depending on your conditions.

All you'll need to do is use an [`updating()` Lifecycle Hook](https://laravel-livewire.com/docs/2.x/lifecycle-hooks) in your component and let the backend handle the logic behind your roles, instead of relying on the frontend (as you'll be more open to injections).

**Reply by mohdaftab** [reply]
@iksaku Thank you so much for the reply.
Someone suggested to use Lifecycle hook and it worked fine but I used updated hook instead of updating.
Do you think updating is the efficient way to use it or updated?
Thank you

`public function updated($name, $value) {
if (preg_match('#manager#i',$value)) {
foreach ($this->role as $rkey=>$role) {
if (!preg_match('#manager#i',$role)) {
unset($this->role[$rkey]);
}
}
$this->role[] = $value;
$this->role = array_unique($this->role);

    } else {
        $this->role = [];
        $this->role[$value] = $value;
    }
    $this->validateOnly($name);
}`

**Reply by iksaku** [reply]
The difference between `updating` and `updated` is that the former is called _before_ assigning the new value, and the latter, being called _after_ the value changes in your component.

If your issue is resolved, remember to mark this discussion as resolved, so people know there's an answer for them here :wink:.

**Reply by mohdaftab** [reply]
Thank you so much
